### PR TITLE
TS-VAD based diarization in CHiME-6

### PR DESCRIPTION
**Pull request by desh2608**
The implementation is from @ipmedenn and it has also been tested by @aarora8 and me.

### CODE REVIEW

File: src/featbin/paste-vectors.cc [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+}
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    using namespace std;
+
+    const char *usage =
+        "Paste vector files \n"
```
Comment (danpovey) [inline src/featbin/paste-vectors.cc:60]: This is not specific: maybe 'Concatenate vectors' might be clearer?

Comment (kkm000) [reply src/featbin/paste-vectors.cc:60]: @danpovey, we have `concat-*` vs. `paste-*` semantics, and this one looks like it's `paste-*`-type? I.e., for
vec1
```
utt-1 [ 1 2 ]
utt-2 [ 3 4 5 ]
```
and vec2
```
utt-1 [ 11 12 ]
utt-2 [ 13 14 15 ]
```
paste (vec1 vec2) is
```
utt-1 [ 1 2 11 12 ]
utt-2 [ 3 4 5 13 14 15 ]
```
Did I grok what the tool was doing?

Comment (danpovey) [reply src/featbin/paste-vectors.cc:60]: From a  UNIX-command-line-manipulating text point of view you're right, but (a) these aren't text files necessarily, (b) someone may not make that mental connection.  These are vectors, "concatenate" is something that makes sense from a vector standpoint.

**Comment by desh2608** [comment]
@kkm000 thanks for the detailed comments! @ipmedenn could you make these changes when you have some time?

**Comment by ipmedenn** [comment]
> @kkm000 thanks for the detailed comments! @ipmedenn could you make these changes when you have some time?

@desh2608 yes I will. Should I make it in kaldi_chime6 repo or in https://github.com/desh2608/kaldi/tree/ts_vad_final?

File: src/featbin/multiply-vectors.cc [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        != kNoRspecifier) {
+      // We're operating on tables, e.g. archives.
+
+      // Last argument is output
+      string wspecifier = po.GetArg(po.NumArgs());
+      BaseFloatVectorWriter vector_writer(wspecifier);
+
+      // First input is sequential
+      string rspecifier1 = po.GetArg(1);
+      SequentialBaseFloatVectorReader input1(rspecifier1);
```
[the remark this answers is not in the document]
Comment (ipmedenn) [reply src/featbin/multiply-vectors.cc:106]: @kkm000 These names have been copypasted from paste-feats.cc, so is it necessary to rename?

File: src/featbin/paste-vectors.cc (hunk 1 above)
Comment (ipmedenn) [reply src/featbin/paste-vectors.cc:60]: @danpovey should I rename paste-vectors to concat-vectors, or you meant only usage string?

**Comment by desh2608** [comment]
> > @kkm000 thanks for the detailed comments! @ipmedenn could you make these changes when you have some time?
> 
> @desh2608 yes I will. Should I make it in kaldi_chime6 repo or in https://github.com/desh2608/kaldi/tree/ts_vad_final?

@ipmedenn You can push it to this branch (ts_vad_final) so that this PR will get updated automatically.

File: src/featbin/multiply-vectors.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (kkm000) [reply src/featbin/multiply-vectors.cc:106]: Yes, please.

File: src/featbin/paste-vectors.cc (hunk 1 above)
Comment (kkm000) [reply src/featbin/paste-vectors.cc:60]: @danpovey, I'm kinda torn on this, the naming would be inconsistent either. We already have binaries with a similar semantics called `paste-*`, and dissimilar called `concat-*`. Maybe `combine-vectors` would work?

Yes, of course I understand these are not usually text files.

**Comment by danpovey** [comment]
I think paste- is fine for the name, for consistency; just change the error
message to clarify what it actually does.

**Comment by ipmedenn** [comment]
> 
> 
> > > @kkm000 thanks for the detailed comments! @ipmedenn could you make these changes when you have some time?
> > 
> > 
> > @desh2608 yes I will. Should I make it in kaldi_chime6 repo or in https://github.com/desh2608/kaldi/tree/ts_vad_final?
> 
> @ipmedenn You can push it to this branch (ts_vad_final) so that this PR will get updated automatically.

@desh2608 it seems that I have no permission to commit in your repo

**Comment by desh2608** [comment]
I believe all the comments have now been addressed by @ipmedenn.

**Comment by ipmedenn** [comment]
@kkm000 check the updated versions, please.

**Comment by kkm000** [comment]
@ipmedenn, Thanks, I'll do my best to review today.

File: src/featbin/paste-vectors.cc [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+}  // namespace kaldi
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    using namespace std;
+
+    const char *usage =
+        "Concatenate vector files \n"
```
Comment (kkm000) [inline src/featbin/paste-vectors.cc:54]: Let's put it this way (feel free to break into lines 60-100 characters on output each paragraph, so they are nice and readable):

>Append to vectors the first archive the vectors under the same key from the second and following archives, and output the result.
>
>NOTE: Keys absent from the first archive will not be present in output, even if they occur in other input archives."

The space before '\n' is not needed.

File: src/featbin/paste-vectors.cc [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    po.Read(argc, argv);
+
+    if (po.NumArgs() < 3) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    if (ClassifyRspecifier(po.GetArg(1), NULL, NULL)
+        != kNoRspecifier) {
+      // We're operating on tables, e.g. archives.
```
Comment (kkm000) [inline src/featbin/paste-vectors.cc:69]: This program can only work on archives. This `if` closes, and the program exists silently if this condition is false.

Reverse the condition and print an error, e.g. 
```suggestion
    if (ClassifyRspecifier(po.GetArg(1), NULL, NULL) != kNoRspecifier) {
      KALDI_ERR << "This program can operate only on tables, not on rxfiles";
    }
```
and then unindent the rest. `KALDI_ERR` will exits the program.

**Comment by ipmedenn** [comment]
@kkm000 Thank you! I've made these changes. @desh2608 could you merge my PR to your repo?

**Comment by kkm000** [comment]
LGTM, thanks guys, merging this!

**Pull request merged**